### ISSUE TITLE
Multisig discussion 

### ISSUE THREAD

**Issue by yuriy77k** [yuriy77k/CallistoBridge#2]
The `owner` should be multisig wallet.

## The issue: who can `transferOwnership`?

Proposal:
 Create create

**Comment by yuriy77k** [comment]
Cons: 
If `founders` can replace the owner they may replace it with an address by their control and get complete control over the Bridge contract. So it will not increase security. The single point of vulnerability will be the `founders` multisig.

**Comment by Dexaran** [comment]
>If founders can replace the owner they may replace it with an address by their control and get complete control over the Bridge contract. So it will not increase security. The single point of vulnerability will be the founders multisig.

The main problem here is that we (1) need the ability to make modifications to the contract i.e. manage authorities and signature rules and (2) we risk compromising keys in the process.

The whole point of `founders` is that it is a different multisig with different implementation and a different set of keys (that are possibly given to different people that are not involved in "owner" multisig). Ideally stored in a different place where "owner" multisig keys are not stored.

Nobody risks compromising `founders` accounts in the process of Bridge setup because these accounts are never involved. They don't sign any transactions, they can rest in cold storages and never be used until its necessary to replace the "owner".

Moreover - `founders` can have different rules of consensus, for example it may be required to have a consensus of 5/6 for founders while a consensus of 3/5 is sufficient for the "owner" multisig. In this case we can retain the ability to act quickly (because 3 persons are easier to communicate) and make necessary workflow changes with "owner" account without a risk of compromising anything.

At the same time a more "hard" consensus of `founders` will be harder to violate.

Stealing 5 keys that are never exposed online and belong to different people is harder than obtaining 3 keys that are sometimes used to sign some transactions.

Why do we need founders at all? - Because if we remove the "owner" and grant all the owner privileges to founders then it effectively turns them into a group of "risky" keys that are sometimes exposed yet a hard consensus of 5/6 is required to perform any update of the contract and thats inconvenient.

**Comment by yuriy77k** [comment]
Agree

**Comment by yuriy77k** [comment]
let's discuss `Founders rules`
1. Founders can't add / remove founder addresses.
2. Only the owner of the Bridge contract can add remove founder once per 24h.
Who can change `founders` wallet in the Bridge contract?

**Comment by Dexaran** [comment]
I proposed to make `founders` addresses implemented directly in the code of the Bridge.
If we decide to use the multisig then it can be permanently hardcoded I think

**Comment by Dexaran** [comment]
We need to attach the code of this multisig in this case and I'm in favor of using a very simplified custom contract instead of utilizing a fully-functional multisig

**Comment by creepas** [comment]
> We need to attach the code of this multisig in this case and I'm in favor of using a very simplified custom contract instead of utilizing a fully-functional multisig

Im also for simplier approach for start as we anyway can do update later on. So why to make it super complicated from the start.

**Comment by yuriy77k** [comment]
I need to know the required majority of founders (in percentage) to approve changes. 

And `Only the owner of the Bridge contract can add remove founder once per 24h.` is it correct? Or will  need to change to 72h

**Comment by Dexaran** [comment]
The timeframe is increased to 72h. Owner can update 1 "founders" account once per 72h

We are going to use a consensus model of 3/5 for "founders"

**Comment by yuriy77k** [comment]
I found an issue in this strategy:

Since `owner` has the right to upgrade the bridge contract, then in case of compromise `owner` the attacker can simply upgrade the bridge with a new contract that gives him full control over it (without any restriction). So does not matter at what time interval `owner` can change `founder` he just removes all of them from the new code.

According this I propose:

**Comment by Dexaran** [comment]
>owner has the right to upgrade the bridge contract

Yes this issue is real but the proposed solution requires a lot of rework. It is much easier to simply restrict the `upgrade` function. 
For example we can place a countdown for 72hours to perform the upgrade.

Or even increase the "setup mode" delay to 72h and only allow to upgrade the contract in "setup mode". This looks like the most suitable solution

**Comment by yuriy77k** [comment]
In this case, we may do like this:
`founders` can upgrade contract but only if it's in "setup mode".
But only `owners` can enable "setup mode".
So to upgrade contrate will require action of `owners` and `founders`

**Comment by creepas** [comment]
I think its solved by setup mode because if owners are compromised we have 72h to vote them out from founders and stop the setup mode or?

**Comment by Dexaran** [comment]
>I think its solved by setup mode because if owners are compromised we have 72h to vote them out from founders and stop the setup mode or?

Yes, it should be solved by `setup mode`.

The proposal:

- Allow upgrades only in `setup mode`

- Enable `setup mode` after 72h since the call of the function.

Only `owners` can call the setup function and only owners can upgrade the contract in the setup mode. Founders serve as a "plan B" and we can vote out the malicious owners if we see someone performing the contract upgrade but it was not a team decision.

**Comment by yuriy77k** [comment]
> we can vote out the malicious owners if we see someone performing the contract upgrade but it was not a team decision.

We could do nothing if will see "contract upgrade" we can see only if "setup mode" is enabled without a team decision.
But enable "setup mode" is a usual operation (every time when we will need to add new pair we have to enable "setup mode" so the attacker can just wait when "setup mode" will be enabled and then upgrade the contract.

Maybe add separate "upgrade mode" with 72h delay, and left "setup mode" with 24h delay for all other settings?

**Comment by Dexaran** [comment]
>Maybe add separate "upgrade mode" with 72h delay, and left "setup mode" with 24h delay for all other settings?

Yes, it is a better option

**Comment by Dexaran** [comment]
It will allow us to keep an eye on a particular event related to the "upgrade mode" and only get notification if that is triggered. Checking every enable of setup mode can be redundant.

**Comment by yuriy77k** [comment]
"upgrade mode" will be not just boolean On/Off, but it will be upgrade request:
```
event UpgradeRequest(address upgrateTo, uint256 validFrom);
```
where: 
- `upgrateTo` - is new contract address
- `validFrom` - timestamp when upgrade maybe perform ( `validFrom` = request timestamp + 3 days)

So we will see the contract address and can check it before upgrate take place

**Comment by Dexaran** [comment]
Looks good to me

**Comment by Dexaran** [comment]
>where:
>upgrateTo - is new contract address
>validFrom - timestamp when upgrade maybe perform ( validFrom = request timestamp + 3 days)

Should we implement an additional function like `confirmUpgrade` that will perform the actual upgrade after the `validFrom` condition is met?

**Comment by yuriy77k** [comment]
no, an actual upgrade will be made by calling function `upgrade()` in proxy

**Comment by yuriy77k** [comment]
https://github.com/yuriy77k/CallistoBridge/blob/2487a3719660c4f7277a1288cc4f73223e052893/contracts/Bridge.sol#L357-L368

https://github.com/yuriy77k/CallistoBridge/blob/2487a3719660c4f7277a1288cc4f73223e052893/contracts/UpgradeableProxyFlattened.sol#L427-L430
